### ISSUE TITLE
Interop issue with range requests when the server sends more bytes than it says

### ISSUE THREAD

**Issue by domenic** [whatwg/fetch#1295]
Source: https://news.ycombinator.com/item?id=28016543

Apparently if the server sends a content-length header of 1024 bytes, but the server instead sends a 4583-byte chunk, Firefox will fail the request and Chrome will continue.

Big providers like Netlify are doing this. Apparently they cannot figure out how to configure their server to send the correct (i.e. post-content-codings) content length: https://answers.netlify.com/t/add-unpacked-content-length-header-for-json-files/1804

(Maybe this should be a HTTPWG issue but I have more faith it'll get sorted out over here given how I've seen issues in that repo go...)

**Comment by domenic** [comment]
Previously noticed in https://github.com/whatwg/fetch/issues/716#issuecomment-788103657, /cc @MattMenke2.

**Comment by MattMenke2** [comment]
I think that's a different issue.  If Chrome sees a content-length for a resource with a chunked transfer encoding, I believe it ignores the content-length.

That comment was about the fact that if we see content-length that matches uncompressed content-length, we'll consider that a success, if decompression results in a file that exactly matches the content-length size.

**Comment by jakearchibald** [comment]
There must be more to this.

Demo: https://github.com/jakearchibald/send-more-bytes-than-length

This has an HTTP/1.1 endpoint that sets a `Content-Length` of 1000 but sends 1500 bytes.

Chrome, Firefox, and Safari all successfully receive the response truncated to 1000 bytes. The server definitely sends all 1500 bytes.

Maybe things are different with chunked encoding, or HTTP/2?

**Comment by MattMenke2** [comment]
That's another issue - due to the linear nature of time, it's not always possible to know if there are more bytes on the wire after you've received 1000 bytes, so browsers don't even try to check for extra bytes.  Also, it's not possible to cancel a read on a socket to check if there are any extra bytes (And, e.g., in the case of SSL, more extra bytes does not mean more extra data), so things can get a bit involved.

**Comment by MattMenke2** [comment]
Also, the extra bytes could be, e.g., in case of latency, a 418 (?) connection timeout response, which isn't actually an error about the previous request.

**Comment by MattMenke2** [comment]
I should add, it's possible to check if there are any bytes currently pending on a socket, and read them if there are, and don't do anything if there aren't, but that's not an operation any browser's TCP logic has previously had to support.  Chrome does have that ability in its socket layer now (with SSL support that doesn't get confused by extra SSL control messages, I believe), but didn't for most of its existence.

As previously noted, of course, that effort is necessarily best-effort, and could get confused by other messages over the wire.

**Comment by MattMenke2** [comment]
Even non-keepalive connections can specify a length, and could theoretically have extra bytes on the wire, which I think corresponds to the case this is trying to consider an error?  Chunked encoding responses could also have extra bytes on the wire (which may not correspond to the issue this is targeted it), and there's the related question of not receiving the terminal when the connection is closed (Chrome treats this as an error, but other browsers do not, last I heard).

**Comment by ronag** [comment]
> and could theoretically have extra bytes on the wire, which I think corresponds to the case this is trying to consider an error

But that at least is possible to detect, i.e. wait for connection close to determine whether extra bytes were sent or not. In the keep-alive case I don't think there is even a theoretical solution to this problem.

**Comment by MattMenke2** [comment]
That may introduce hangs in the case of currently cross-browser functional servers that expect browsers to close non-keepalive connections that specify lengths (which may be an uncommon enough case that it's fine to do, but is definitely a potentially breaking change)

**Comment by domenic** [comment]
https://phiresky.netlify.app/blog/2021/hosting-sqlite-databases-on-github-pages/ still gives an error in Firefox and no error in Chrome, so there's something going on here, although maybe it's not what the HackerNews crowd diagnosed.

I spent 60 seconds trying to debug this but the code is obfuscated and not that easy. The network panel on Firefox displays this info which might be interesting:

![image](https://user-images.githubusercontent.com/617481/129593888-58c8b3bc-fdbb-43db-be0a-91791086c87d.png)

as reported, the server sends a good bit more than the requested 1024 bytes, even though it says it's sending "bytes 0-1023/1247079".

If this isn't enough of a clue for people then we can probably close this or maybe transform it into a Gecko bug to see if their engineers can gather more insight.

**Comment by MattMenke2** [comment]
That response is a 206 with a content-encoding - the request should be treated as a failure because of that, if for no other reason.

**Comment by domenic** [comment]
Interesting. And in Chrome they don't send content-encoding:

![image](https://user-images.githubusercontent.com/617481/129596136-2cfefdcb-cf60-43c1-a727-3752013d7368.png)

(although they do send it on the initial 200)

Either Firefox's inspector is carrying over headers from the initial 200 to the 206 (seems unlikely), or this is a website bug where they're changing their behavior based on the browser in question.

**Comment by ddragana** [comment]
I tried o reproduce the issue, but I am getting a different error:
"Length of the file not known. It must either be supplied in the config or given by the HTTP server."
And there is an warning before:
"Warning: The server responded with br encoding to a HEAD request. Ignoring since it may not do so for Range HTTP requests, but this will lead to incorrect results otherwise since the ranges will be based on the compressed data instead of the uncompressed data."

Firefox receives a 206 with a content-encoding. The header is coming from the server. I cannot confirm that that was causing the original issue because I cannot reproduce the error.
